test/test_decomp.py

```diff
@@ -0,0 +1,458 @@
+        (torch.float32, torch.ops.aten.native_layer_norm_backward.default): (
+            1e-3,
+            1e-3,
+        ),
+    }
+    if (b.dtype, op) in tol_table:
+        rtol, atol = tol_table[(b.dtype, op)]
+    else:
+        rtol, atol = _getDefaultRtolAndAtol(a.dtype, b.dtype)
+    test_case.assertEqual(a, b, rtol=rtol, atol=atol)
+
+
+# Given f, returns an f' such that:
+# - f' takes only positional arguments
+# - All arguments to f' are floating-point Tensors
+# - All outputs of f' are floating-point Tensors
+def normalize_op_input_output2(
+    f, args, kwargs, output_process_fn_grad=None, requires_grad=True
+):
+    flat_args, args_spec = tree_flatten(args)
+    diff_argnums = tuple(
+        i
+        for i, arg in enumerate(flat_args)
+        if diff_arg(arg, requires_grad=requires_grad)
+    )
+    assert len(diff_argnums) > 0
+    primals = tuple(flat_args[i] for i in diff_argnums)
+
+    @functools.wraps(f)
+    def wrapped(*primals):
+        _args = list(flat_args)
+        for num, arg in zip(diff_argnums, primals):
+            _args[num] = arg
+        _args = tree_unflatten(_args, args_spec)
+        result = f(*_args, **kwargs)
+        if output_process_fn_grad is not None:
+            result = output_process_fn_grad(result)
+        if isinstance(result, tuple):
+            # TODO: Remove the following hack for namedtuples
+            result = tuple(result)
+            result = tuple(
+                r
+                for r in result
+                if isinstance(r, Tensor) and (r.is_floating_point() or r.is_complex())
+            )
+            assert len(result) > 0
+        return result
+
+    return wrapped, primals
+
+
+# NB: This also upcasts dtype arguments
+
+
+def upcast_tensor(func, x, dtype=torch.float32):
+    # TODO: stop hardcoding integer values to pass in
+    # dtype in torch.ops
+    FLOAT16_DTYPE = 5
+    BFLOAT16_DTYPE = 15
+    FLOAT64_DTYPE = 7
+
+    # Some functions take a dtype as argument, so we need to
+    # manually change that dtype in order to run it with a
+    # higher precision
+    dtype_arg_table = {
+        torch.ops.aten._softmax_backward_data.default,
+        torch.ops.aten._log_softmax_backward_data.default,
+    }
+
+    if isinstance(x, Tensor) and x.dtype.is_floating_point:
+        return x.to(dtype=dtype)
+    elif (
+        isinstance(x, int)
+        and func in dtype_arg_table
+        and x in [FLOAT16_DTYPE, BFLOAT16_DTYPE]
+    ):
+        return FLOAT64_DTYPE
+    else:
+        return x
+
+
+def normalize_op_input_output(f, sample, requires_grad=True):
+    args = tuple([sample.input] + list(sample.args))
+    return normalize_op_input_output2(
+        f,
+        args,
+        sample.kwargs,
+        sample.output_process_fn_grad,
+        requires_grad=requires_grad,
+    )
+
+
+CROSS_REF_EXCLUDE_SET = {
+    (
+        "cpu",
+        torch.bfloat16,
+        "nn.functional.layer_norm",
+    ),  # "batch_norm" not implemented for 'BFloat16'
```

Comment:
Why is this needed - is it not already covered by OpInfo dtype support?

Reply:
I am not sure but empirically the test failed even after I made us respect backward dtypes.